fix(ScrollAction): unsafe non-null assertion

kszczek opened the pull request:
This conditional most likely was meant to use `||` instead of `&&` operator to check if `notificationContext` is null OR if no Scrollable was found for this `notificationContext`. Using the AND operator instead causes the null assertion to fail when notificationContext is null, while also causing an assertion failure when notificationContext is not null, but no Scrollable is found for it. While at it, let's also deduplicate the Scrollable.maybeOf call.

This fix is purely speculative, I didn't encounter this error, but rather was reading the ScrollAction's source and stumbled upon this. ~~With that said I wasn't actually able to trigger this error condition, thus why there are no tests included with this PR, if anyone has ideas on how this could be triggered, I'll be happy to include such test cases.~~

fixes #158063

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

kszczek committed:
fix(ScrollAction): unsafe non-null assertion

This conditional most likely was meant to use || instead of && operator
to check if notificationContext is null OR if no Scrollable was found
for this notificationContext. Using the AND operator instead causes the
null assertion to fail when notificationContext is null, while also
causing an assertion failure when notificationContext is not null, but
no Scrollable is found for it. While at it, let's also deduplicate the
Scrollable.maybeOf call and remove the now unnecessary assertion.

Fixes: https://github.com/flutter/flutter/issues/158063

nate-thegrate commented on packages/flutter/lib/src/widgets/scrollable_helpers.dart:
```diff
@@ -460,17 +460,17 @@ class ScrollAction extends ContextAction<ScrollIntent> {
       }
-      state = Scrollable.maybeOf(primaryScrollController.position.context.notificationContext!);
     }
-    assert(state != null, '$ScrollAction was invoked on a context that has no scrollable parent');
```
I agree it's a bit strange for this method to use a combination of `assert`s and early `return`s, rather than sticking with one or the other.

kszczek replied:
Yeah, this assert was also marked as unnecessary by the linter after applying the patch.

nate-thegrate commented on packages/flutter/lib/src/widgets/scrollable_helpers.dart:
```diff
@@ -460,17 +460,17 @@ class ScrollAction extends ContextAction<ScrollIntent> {
 
-      if (primaryScrollController.position.context.notificationContext == null
-          && Scrollable.maybeOf(primaryScrollController.position.context.notificationContext!) == null) {
+      if (primaryScrollController.position.context.notificationContext != null) {
+        state = Scrollable.maybeOf(primaryScrollController.position.context.notificationContext!);
+      }
```
I think it'd be good to set a local variable here, to further reduce the amount of null-asserting.

```suggestion
      final BuildContext? notificationContext = primaryScrollController.position.context.notificationContext;
      if (notificationContext != null) {
        state = Scrollable.maybeOf(notificationContext);
      }
```

kszczek replied:
Done.

nate-thegrate reviewed (commented):
Looks great!

Just 1 thing left: go to [scrollable_helpers_test.dart](https://github.com/kszczek/flutter/blob/fix-scroll-action-condition/packages/flutter/test/widgets/scrollable_helpers_test.dart) and add a test with something like
```dart
somethingThatShouldWork();
expect(tester.takeException(), isNull);
```

kszczek commented:
Thanks for the review, I've addressed the code style issue and added the aforementioned test.

I've had to mock the `Scrollable.notificationContext` property though, because I couldn't find a way to make it null when the action is invoked. I don't think that's practically possible, as warranted by the fact that no user reported this issue.

nate-thegrate commented on packages/flutter/test/widgets/scrollable_helpers_test.dart:
```diff
@@ -590,4 +606,77 @@ void main() {
+    );
+
+    for (final LogicalKeyboardKey key in keys) {
+      // We exclude the modifier keys here for web testing since default web shortcuts
+      // do not use a modifier key with arrow keys for ScrollActions.
```
```suggestion
      // The default web shortcuts do not use a modifier key for ScrollActions.
```

nate-thegrate commented on packages/flutter/test/widgets/scrollable_helpers_test.dart:
```diff
@@ -590,4 +606,77 @@ void main() {
+        await tester.sendKeyDownEvent(modifierKey);
+      }
+
+      // Invoking the ScrollAction would previously lead to a type error caused by an
+      // unsafe non-null assertion on the notificationContext, let's make sure it
+      // doesn't happen anymore.
```
Instead of describing the bug in the test code, we usually just add an issue link.

```suggestion
```

nate-thegrate commented on packages/flutter/test/widgets/scrollable_helpers_test.dart:
```diff
@@ -590,4 +606,77 @@ void main() {
     expect(find.text('The cape as red as blood'), findsOneWidget);
     expect(find.text('The hair as yellow as corn'), findsOneWidget);
   });
+
+  testWidgets('Invoking a ScrollAction when notificationContext is null does not cause an exception.', (WidgetTester tester) async {
```
```suggestion

  // Regression test for https://github.com/flutter/flutter/issues/158063.
  testWidgets('Invoking a ScrollAction when notificationContext is null does not cause an exception.', (WidgetTester tester) async {
```

I went ahead and made an issue so we can link it here:
- https://github.com/flutter/flutter/issues/158063

nate-thegrate commented on packages/flutter/test/widgets/scrollable_helpers_test.dart:
```diff
@@ -590,4 +606,77 @@ void main() {
+        await tester.sendKeyUpEvent(modifierKey);
+      }
+
+      // We expect the ScrollAction to not find any eligible scrollables, thus the
+      // scroll position should not change.
```
```suggestion
      // No scrollable is found, so the scroll position should not change.
```

nate-thegrate reviewed (approved):
LGTM, thanks for the well-structured test!

Perhaps if it's impossible for the `notificationContext` to be null, the getter should be updated to be non-nullable, but that's outside the scope of this PR.

<br>

I have a few tiny suggestions for the test comments (for brevity and to follow [style guidelines](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-the-passive-voice-recommend-do-not-require-never-say-things-are-simple)), so take a look when you get a chance :)

kszczek commented:
All done!

nate-thegrate commented:
Awesome, thanks @kszczek! We'll be able to merge once the PR has 2 approvals.

Piinks reviewed (commented):
This LGTM. I was a little wary that we did not have an issue demonstrating this causing a problem, but it looks like a nice clean up.

From an offline chat with @nate-thegrate, the notification context can be null, it comes from the key for the gesture detector in ScrollableState: 
https://github.com/flutter/flutter/blob/793569f621f79119dd7cb2c82c403b5d52271946/packages/flutter/lib/src/widgets/scrollable.dart#L598
